Extend image button to insert external images

duarteframos opened the pull request:
This basically reuses mostly code from https://github.com/Jermolene/TiddlyWiki5/blob/master/core/ui/EditorToolbar/link-dropdown.tid

It adds an edit text box with an insert button that allows inserting external images in wikitext through their URL.

Please check code sanity, it may be over complicated or have redundant parts, I am not at all a coder. It may also need new localized language strings

Ideally this text box would also allow filtering or searching for internal image tiddlers displayed in the list bellow, just like in the links button dropdown currently, but that is way above my skill level.
It should also commit changes when pressing enter key

Discussion here https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/tiddlywiki/7LfTfO7oC9U/lT_P5mi4EQAJ

Commit:
Update picture-dropdown.tid

tobibeer commented on core/ui/EditorToolbar/picture-dropdown.tid:
```diff
@@ -4,8 +4,44 @@ title: $:/core/ui/EditorToolbar/picture-dropdown
 [img[$(imageTitle)$]]
 \end
 
+\define insert-external-image()
```
As said, not sure if we're really aiming at external images here. I would think the extension should rather render the html for embedding image files outside of tiddlywiki, rather than to enter some title as if that was an image tiddler.

The thing is, we already know what image tiddlers are there, and that's what the current implementation relies on.

So, to me it would rather be helpful to filter upon those images via search terms..

tobibeer replied:
Sry, Duarte, double posting: as you can [read here](https://groups.google.com/d/msg/tiddlywiki/7LfTfO7oC9U/OkH7Zz--EQAJ) I had quite a weird misunderstanding about the ways the **[img[]]** markup works.

tobibeer replied:
> Ideally nothing would show up initially until the user typed something, since if there are lots of internal images, loading them all for displaying is quite slow and locks up the browser.

I think rather than not display anything, it would be better to limit the results, e.g. to 50 or 100.

> For external images tiddlywiki would try to display a small preview/thumbnail in the dropdown of the resulting path (as you suggested) confirming it was valid

Cool idea and possibly rather simple: just preview the output as it would look.

duarteframos commented on core/ui/EditorToolbar/picture-dropdown.tid:
```diff
@@ -4,8 +4,44 @@ title: $:/core/ui/EditorToolbar/picture-dropdown
 [img[$(imageTitle)$]]
 \end
 
+\define insert-external-image()
```
OK, fair enough, just though it might be useful for someone else since it fits my use case though it certainly could be improved as you suggested by displaying the image preview.

Feel free to close this anytime, and thanks for the feedback

tobibeer replied:
@duarteframos, feel free to delete your comment. It was all my misunderstanding. ;-)

duarteframos replied:
No problem, as mentioned before, I am not a coder by any stretch of the word, just a bare copy-paster at tops. I hacked this together, but like you said this could be tremendously improved.
The way I see it it would be perfect if input text in the field would:

1. Filter the displayed results for image tiddlers.

2. Ideally nothing would show up initially until the user typed something, since if there are lots of internal images, loading them all for displaying is quite slow and locks up the browser.

3. This would allow inserting external images with an protocol+UTL (like http://path.to/image/file.jpg), or relative path (like media/path/to/image.jpg)

3. For external images tiddlywiki would try to display a small preview/thumbnail in the dropdown of the resulting path (as you suggested) confirming it was valid

2. Like the [proposal I did before for the links dropdown](https://github.com/Jermolene/TiddlyWiki5/issues/2521) this would also allow inserting not-yet existing image tiddlers

4. Pressing enter would commit whatever was introduced into the text box, internal image tiddler or not

Well at least that is what I would do, if I had the skills to accomplish it ;) I was in the hopes that Jeremy, or another talented developer such as yourself would improve it based on this concept or starting from scratch.

tobibeer replied:
> Ideally nothing would show up initially until the user typed something, since if there are lots of internal images, loading them all for displaying is quite slow and locks up the browser.

I think rather than not display anything, it would be better to limit the results, e.g. to 50 or 100.

> For external images tiddlywiki would try to display a small preview/thumbnail in the dropdown of the resulting path (as you suggested) confirming it was valid

Cool idea and possibly rather simple: just preview the output as it would look.

duarteframos replied:
Limiting the displayed thumbnails number seems like a good compromise.
Not only before typing (maybe sorted by creation date, most recent displayed only) but also for search result matches after the user types something (sorted by relevance).

Perhaps even a lower number around the twenties is advisable for lower end devices. 

Thanks for the feedback, and taking interest in this.

tobibeer replied:
Made a more comprehensive demo here:

http://insert-image.tiddlyspot.com

tobibeer replied:
Feel free to pick from it for this PR.

duarteframos replied:
Oh man that looks awesome, seems to be doing everything I envisioned.

I'll give it a more thorough look later tonight when I'm back at my desktop.
Thanks for all the hard work :)

duarteframos replied:
OK, tested on my desktop browser, works like a charm, seems like a much needed improvement thanks for all the improvements.
So what's the plan to get this into official release? I no longer feel like this should be in my PR, it is after all your code skill here, not mine. :)

tobibeer replied:
As you wish. I really don't mind.

If you have the time to give it a final polish and do a PR, please do:

1. styles probably need to go into vanilla/base
2. the buttons probably need tooltips (similar to AdvancedSearch)
3. the PR should point to a demo to show what changed / how it works

duarteframos replied:
Alright, I'll update this PR later today

Commit:
Editor Toolbar Picture Dropdown Improvements

Code is actually from Tobias Beer, improvements as described in the [working example at tiddlyspot](http://insert-image.tiddlyspot.com/)
Adds a search field to the Editor Toolbar Picture Dropdown tiddler capable of both filtering internal images and/or inserting external images from any URL

Commit:
Style improvements for picture dropdow

Style improvements to accommodate changes for the the picture dropdown
Code by Tobias Beer as described in the [working example at TiddlySpot](http://insert-image.tiddlyspot.com/)

Commit:
Improvements for the Image Picker

Changes by Tobias Beer as described at the[ working example in TiddlySpot](http://insert-image.tiddlyspot.com/) to accommodate improvements to the picture dropdown for the image editor

This allows displaying previews for external images before inserting in wikitext

duarteframos commented:
Alright, added your code changes to the branch, hope everything is correct, not very versed with how GIT works. [Working demo here](http://insert-image.tiddlyspot.com/)
